instrumentation/jms/jms-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jms/MessageWithDestination.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.jms;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class MessageWithDestination {
+
+  // visible for tests
+  static final String TIBCO_TMP_PREFIX = "$TMP$";
+
+  public abstract MessageAdapter message();
+
+  public abstract String destinationName();
+
+  public abstract String destinationKind();
+
+  public abstract boolean isTemporaryDestination();
+
+  public static MessageWithDestination create(
+      MessageAdapter message, DestinationAdapter fallbackDestination) {
+    DestinationAdapter jmsDestination = null;
+    try {
+      jmsDestination = message.getJmsDestination();
+    } catch (Exception ignored) {
+      // Ignore
+    }
+    if (jmsDestination == null) {
+      jmsDestination = fallbackDestination;
+    }
+
+    if (jmsDestination.isQueue()) {
+      return createMessageWithQueue(message, jmsDestination);
+    }
+    if (jmsDestination.isTopic()) {
+      return createMessageWithTopic(message, jmsDestination);
+    }
+    return new AutoValue_MessageWithDestination(
+        message, "unknown", "unknown", /* isTemporaryDestination= */ false);
+  }
+
+  private static MessageWithDestination createMessageWithQueue(
+      MessageAdapter message, DestinationAdapter queue) {
+    String queueName;
+    try {
+      queueName = queue.getQueueName();
+    } catch (Exception e) {
+      queueName = "unknown";
+    }
+
+    boolean temporary = queue.isTemporaryQueue() || queueName.startsWith(TIBCO_TMP_PREFIX);
+
+    return new AutoValue_MessageWithDestination(message, queueName, "queue", temporary);
+  }
+
+  private static MessageWithDestination createMessageWithTopic(
+      MessageAdapter message, DestinationAdapter topic) {
+    String topicName;
+    try {
+      topicName = topic.getTopicName();
+    } catch (Exception e) {
+      topicName = "unknown";
+    }
+
+    boolean temporary = topic.isTemporaryTopic() || topicName.startsWith(TIBCO_TMP_PREFIX);
```

Comment:
Does it make sense to try and encapsulate the name check in the `DestinationAdapter` implementation? It's a bit of a weak api right now, because the implementation only gives part of the answer (eg. it can lie to the user).

Reply:
The whole purpose of the `DestinationAdapter` is just to hide the JMS API, not to include some arbitrary name-based logic. I agree that this looks kinda bad in the `MessageWithDestination`, but I wouldn't like duplicating this arcane TIBCO_TMP logic in both adapter implementations either.